crypto/src/merkle/mod.rs

```diff
@@ -216,8 +215,8 @@ impl<H: Hasher> MerkleTree<H> {
         }
 
         let index_map = map_indexes(indexes, self.depth())?;
+        let mut leaves = vec![H::Digest::default(); indexes.len()];
```

Comment:
Wouldn't this create a bigger vector than we may need? Basically, if there are duplicate indexes, we'd allocate more a bigger vector than the number of leaves we'll eventually have. Or am I missing something?

Reply:
That's correct and without that we would have an out-of-range index error here https://github.com/facebook/winterfell/blob/0b79eaef9b561995f750e123edec419fab83f5ea/crypto/src/merkle/mod.rs#L230

Reply:
I looked into this a bit more, and while this works, I'm not sure this is the right solution. One of the issues is that contrary to what I said above `map_indexes()` function behaves differently for inputs with and without duplicates. But if we fix `map_indexes()` the verification procedure starts failing because of how we serialize/deserialize proofs. Specifically, when we serialize a proof, we always assume that the number of openings is the same as the number of queries specified by the protocol. So, we must serialize extra "dummy" values so that they can be deserialized properly. This is why this change works.

I think the proper solution is to update serialization code so that it serializes the actual number of openings. But this is a bigger change. I'll need to think about the best way to implement it.

Reply:
I see, my question is then why not just deduplicate the indices at the start of `prove_batch`?

Reply:
This seems to work only if we don't de-duplicate indexes in [BatchMerkleProof::get_root()](https://github.com/facebook/winterfell/blob/main/crypto/src/merkle/proofs.rs#L123) method. In theory, we should do this as well - but then the tests stop passing. So, I think this approach is fragile.

I've proposed an alternative approach in #224. Let me know what you think.